### PR TITLE
Add cuda device synchronization support in JIT

### PR DESCRIPTION
**Pull request by nikithamalgifb**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55469 Add cuda device synchronization support in JIT**

Differential Revision: [D27749077](https://our.internmc.facebook.com/intern/diff/D27749077)

### CODE REVIEW

File: test/jit/test_cuda.py [hunk 1]
```diff
@@ -49,6 +49,35 @@ def tearDown(self):
...
+        def test_device_synchronize():
+            prev_current_device_index = torch.cuda.current_device()
+            torch.cuda.synchronize()
+            torch.cuda.synchronize('cuda')
+            torch.cuda.synchronize('cuda:0')
+            torch.cuda.synchronize(0)
+            torch.cuda.synchronize(torch.device('cuda:1'))
+            after_current_device_index = torch.cuda.current_device()
+
+            return prev_current_device_index == after_current_device_index
```
Comment (SplitInfinity) [inline test/jit/test_cuda.py:67]: Consider adding a comment explaining why this check is needed.

File: test/jit/test_cuda.py [hunk 2]
```diff
@@ -49,6 +49,35 @@ def tearDown(self):
...
+
+        @torch.jit.script
+        def test_multi_device_synchronize():
+            torch.cuda.synchronize(torch.device('cuda:0'))
+            prev_current_device_index = torch.cuda.current_device()
+            torch.cuda.synchronize(1)
+            after_current_device_index = torch.cuda.current_device()
+
+            return prev_current_device_index == after_current_device_index
```
Comment (SplitInfinity) [inline test/jit/test_cuda.py:76]: Consider adding a comment explaining why this check is needed.

File: torch/csrc/jit/runtime/register_cuda_ops.cpp [hunk 3]
```diff
@@ -119,6 +119,54 @@ RegisterOperators const reg({
...
+          // to synchronize.
+          if (current_device_index != device_index) {
+            c10::cuda::set_device(device_index);
+          }
+          c10::cuda::device_synchronize();
+
+          // Reset the device to current_device before synchronizing.
+          if (current_device_index != device_index) {
+            c10::cuda::set_device(current_device_index);
+          }
```
Comment (SplitInfinity) [inline torch/csrc/jit/runtime/register_cuda_ops.cpp:145]: This code looks extremely similar to the implementation for the `int64_t` variant below, consider refactoring.

**Review by SplitInfinity — COMMENTED** [review]
Do we need to mark these operators as having side effects so that they don't get DCE'd? 

Consider using FileCheck in the tests to make assertions about what's in the graph. I wouldn't be surprised if the tests in their current form aren't actually calling the ops...

**Comment by nikithamalgifb** [comment]
> Do we need to mark these operators as having side effects so that they don't get DCE'd?
> 
> Consider using FileCheck in the tests to make assertions about what's in the graph. I wouldn't be surprised if the tests in their current form aren't actually calling the ops...

Yeah, they were getting DCE'd. I have fixed this now.

File: torch/csrc/jit/runtime/register_cuda_ops.cpp (hunk 3 above)
Comment (nikithamalgifb) [reply torch/csrc/jit/runtime/register_cuda_ops.cpp]: Done

File: test/jit/test_cuda.py (hunk 2 above)
Comment (nikithamalgifb) [reply test/jit/test_cuda.py:81]: Done

**Review by SplitInfinity — APPROVED** [review]
Thanks for addressing the comments I made. I would still highly recommend adding some FileCheck checks to the unit tests to make sure the operator isn't getting DCE'd.

**Comment by nikithamalgifb** [comment]
> Thanks for addressing the comments I made. I would still highly recommend adding some FileCheck checks to the unit tests to make sure the operator isn't getting DCE'd.

Thanks Meghan. I have added the checks to synchronize test cases. I'll create a new PR to handle FileChecks for the rest of the tests

**Comment by SplitInfinity** [comment]
> > Thanks for addressing the comments I made. I would still highly recommend adding some FileCheck checks to the unit tests to make sure the operator isn't getting DCE'd.
> 
> Thanks. I have added the checks to synchronize test cases. I'll create a new PR to handle FileChecks for the rest of the tests

Oops, I did not see that. Sorry!

**Pull request closed without merging**